src/V3Scope.cpp

```diff
@@ -261,7 +266,18 @@ class ScopeVisitor final : public VNVisitor {
     void visit(AstVar* nodep) override {
         // Make new scope variable
         if (!nodep->user1p()) {
-            AstVarScope* const varscp = new AstVarScope(nodep->fileline(), m_scopep, nodep);
+            AstScope* scopep = m_scopep;
+            if (AstIfaceRefDType* const ifacerefp = VN_CAST(nodep->dtypep(), IfaceRefDType)) {
+                // Attach every non-virtual interface variable its inner scope
+                if (ifacerefp->cellp() && !ifacerefp->cellp()->isVirtIface())
+                    scopep = VN_AS(ifacerefp->cellp()->user2p(), Scope);
+            }
+
+            AstVarScope* const varscp = new AstVarScope{nodep->fileline(), scopep, nodep};
+            if (m_aboveCellp && m_aboveCellp->isVirtIface()) {
+                // Prevent optimizations to this variable: may be cross-accessed
+                nodep->sigPublic(true);
```

Comment:
Make a new function e.g. sigVirtIFRef or something, and use that in most places where sigPublic is tested.  But perhaps not all.